### PR TITLE
docs: add tests for lazy loading AngularJS example

### PR DESCRIPTION
**Pull request by brandonroberts**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [x] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: aio/tools/examples/shared/package.json [hunk 1]
```diff
@@ -32,7 +32,9 @@
     "@nguniversal/common": "8.0.0-rc.0",
     "@nguniversal/express-engine": "8.0.0-rc.0",
     "@nguniversal/module-map-ngfactory-loader": "^8.0.0-rc.0",
+    "angular": "1.7.7",
```
Comment (gkalpak) [inline aio/tools/examples/shared/package.json]: Why not 1.7.8? 🤔

File: aio/tools/examples/shared/boilerplate/cli-ajs/package.json [hunk 2]
```diff
@@ -0,0 +1,56 @@
+{
```
Comment (gkalpak) [inline aio/tools/examples/shared/boilerplate/cli-ajs/package.json]: More boilerplate. Yay! 😛

File: aio/content/examples/upgrade-lazy-load-ajs/src/app/lazy-loader.service.ts [hunk 3]
```diff
@@ -4,17 +4,10 @@ import { Injectable } from '@angular/core';
...
-
   load(el: HTMLElement): void {
-    if (this.bootstrapped) {
-      return;
-    }
-
     import('./angularjs-app').then(app => {
       try {
         app.bootstrap(el);
-        this.bootstrapped = true;
```
Comment (gkalpak) [inline aio/content/examples/upgrade-lazy-load-ajs/src/app/lazy-loader.service.ts]: Why these changes?

File: aio/content/examples/upgrade-lazy-load-ajs/example-config.json [hunk 4]
```diff
@@ -0,0 +1,3 @@
+{
+    "projectType": "cli-ajs"
```
Comment (gkalpak) [inline aio/content/examples/upgrade-lazy-load-ajs/example-config.json]: 4 spaces? 🤔

File: aio/content/examples/upgrade-lazy-load-ajs/e2e/src/app.e2e-spec.ts [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+            // TODO: must manually assign __TESTABILITY__NG1_APP_ROOT_INJECTOR__ (https://github.com/angular/angular/issues/22723)
+            `window.__TESTABILITY__NG1_APP_ROOT_INJECTOR__ = angular.resumeBootstrap(arguments[0]) `
+            + `|| angular.element('app-angular-js').injector();`,
+            'resume bootstrap',
+            moduleNames
+        );
+    }
+
+    // Wait for the initial AngularJS page to finish loading
+    await browser.waitForAngularEnabled(true);
```
Comment (gkalpak) [inline aio/content/examples/upgrade-lazy-load-ajs/e2e/src/app.e2e-spec.ts]: Why is this needed? Isn't this on by default?

File: aio/content/examples/upgrade-lazy-load-ajs/e2e/src/app.e2e-spec.ts [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+  it('should display \'Angular 404\' when visiting an invalid URL', async() => {
+    await pageElements.notFoundPageHref.click();
+
+    const paragraphText = await pageElements.notFoundPageParagraph.getText();
+
+    expect(paragraphText).toEqual('Angular 404');
+  });
+
+  // Workaround for https://github.com/angular/protractor/issues/4724
+  async function loadAngularJS() {
```
Comment (gkalpak) [inline aio/content/examples/upgrade-lazy-load-ajs/e2e/src/app.e2e-spec.ts]: Oh my 😱 😱 😱
Having to have this helper function looks terrible :grin:

We should either ship it as an ngUpgrade helper or (even better) as a Protractor helper (assuming that Protractor cannot be fixed to automatically detect this).

File: aio/tools/examples/shared/boilerplate/cli-ajs/package.json (hunk 2 above)
Comment (brandonroberts) [reply aio/tools/examples/shared/boilerplate/cli-ajs/package.json]: 😂 I know ... Refactoring this infrastructure is on the list 😛

File: aio/content/examples/upgrade-lazy-load-ajs/src/app/lazy-loader.service.ts (hunk 3 above)
Comment (brandonroberts) [reply aio/content/examples/upgrade-lazy-load-ajs/src/app/lazy-loader.service.ts]: The router destroys the component that holds the `ng-view`, so the options are bootstrap every time, or add a custom `RouteReuseStrategy` to retain the component and the root element. @jasonaden said bootstrapping was the recommendation for now.

File: aio/content/examples/upgrade-lazy-load-ajs/e2e/src/app.e2e-spec.ts (hunk 5 above)
Comment (brandonroberts) [reply aio/content/examples/upgrade-lazy-load-ajs/e2e/src/app.e2e-spec.ts]: This snippet was brought over from https://github.com/angular/protractor/issues/4724 that reported the same issue. I'll remove it and see if it still works

File: aio/content/examples/upgrade-lazy-load-ajs/e2e/src/app.e2e-spec.ts (hunk 6 above)
Comment (brandonroberts) [reply aio/content/examples/upgrade-lazy-load-ajs/e2e/src/app.e2e-spec.ts]: Agreed, that's why I left the note about the workaround, because its been a known issue for a while that hasn't been addressed yet.

File: aio/content/examples/upgrade-lazy-load-ajs/src/app/lazy-loader.service.ts (hunk 3 above)
Comment (gkalpak) [reply aio/content/examples/upgrade-lazy-load-ajs/src/app/lazy-loader.service.ts]: That is definitely _NOT_ recommended by me :grin:
Main reasons:

a. Depending on how you do the bundling, you may end up loading and running the same code repeatedly (and unnecessarily).
    This doesn't seem to be an issue with how cli+webpack work at the time (AngularJS is only loaded/parsed once).
b. Since you are not cleaning up (e.g. by means of `$rootScope.$destroy()`), a bunch of listeners (e.g. on `document` or `window`) are left behind. In addition to that, these listeners may hold on to other stuff, leading to memory leaks.
c. As a result of the above, AngularJS state is retained, which could lead to buggy (and really hard to debug) behavior, esp. if you have multiple AngularJS routes.

Let's discuss this offline.

**Review by gkalpak — CHANGES_REQUESTED** [review]

**Review by gkalpak — APPROVED** [review]
Awesome! We now have a cli-based hybrid app and we have tests to prove it works :tada:
